linkerd/stack/tracing/src/lib.rs

```diff
@@ -148,23 +108,26 @@ where
     type Future = Instrumented<P::Future>;
 
     fn proxy(&self, svc: &mut S, request: Req) -> Self::Future {
-        let _enter = self.span.enter();
+        let span = self.get_span();
+        let _enter = span.enter();
         trace!(?request, "proxy");
-        self.inner.proxy(svc, request).instrument(self.span.clone())
+        self.inner.proxy(svc, request).instrument(span.clone())
     }
 }
 
-impl<Req, S> tower::Service<Req> for Instrument<S>
+impl<Req, T, G, S> tower::Service<Req> for Instrument<T, G, S>
 where
     Req: std::fmt::Debug,
+    G: GetSpan<T>,
     S: tower::Service<Req>,
 {
     type Response = S::Response;
     type Error = S::Error;
     type Future = Instrumented<S::Future>;
 
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        let _enter = self.span.enter();
+        let span = self.get_span();
```

Comment:
just a thinky: creating a new separate span for `poll_ready` kiiiiinda makes me wonder if we want that span to _indicate_ that it's a `poll_ready`...but that would be a bigger change to make the `GetSpan` stuff aware of `poll_ready` calls somehow, or to call `get_span` and then immediately create a new child of that span for the `poll_ready`. so maybe let's punt on that for now...

Reply:
yet another idle thinky is that _maybe_ the most semantically correct thing is to have one span that covers poll_ready and the next call after that poll_ready, like
```rust
struct InstrumentedService<S, G> {
    span: Option<Span>,
    get_span: G,
    // ...
}

impl<S, G, R> Service<R> for InstrumentedService<S, G> 
where
    // ...
{
    // ...
    
    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
         let _enter = self.span
            .get_or_insert_with(|| self.get_span(self.target))
            .enter();
            
        // ...
    }

    fn call(&mut self, req: Request) -> Self::Future {
        let span = self.span
            .take()
            .unwrap_or_else(|| self.get_span(self.target))
            .entered();

       trace!(?request, "service");
       self.inner.call(request).instrument(span.exit())
    }
}
```
that way, you have one span whose total active time is the sum of the time spent polling the service, calling it, and polling its future, and whose total idle time includes the time between when the service was driven to readiness and when it was called. but, this maybe only matters if we are actually timing spans, rather than just using them for log contexts.

also, this means that the future's span's parent is the current span when the service was _polled_, which may not always be the same span the service was _called_ in...

anyway, just thinking out loud here.

Reply:
I made this change in ad87c97.